-i flag for `tsh ssh` implementation.

kontsevoy opened the pull request:
This PR implements #1033

kontsevoy committed:
Minor chnges to tctl (langauge/naming)

This commit does not change tctl behavior

kontsevoy committed:
Merge remote-tracking branch 'origin/master' into ev/i

kontsevoy committed:
Removed "upserted" language from tctl CLI

Fixes #857

kontsevoy commented:
retest this please

kontsevoy committed:
Added a bit more logging to session creation code

kontsevoy committed:
Implemented identity loading

First part of addressing #1033 is ability to load credentials from the
credentials file(s).

This commit adds -i flag processing, i.e. a certificate can be fed via a
cert.file and used to login.

kontsevoy committed:
Added host authorization to identity-based auth

Hooked up -i auth into the proper place

kontsevoy committed:
Covered host authentication with tests

kontsevoy committed:
Merge branch 'master' into ev/i

kontsevoy commented:
retest this please

russjones commented on lib/client/keyagent.go:
```diff
@@ -240,12 +267,19 @@ func (a *LocalKeyAgent) CheckHostSignature(hostId string, remote net.Addr, key s
 	}
-	err = trace.AccessDenied("untrusted host %v", hostId)
-	log.Error(err)
+	// final step: lets ask user:
```
Can you update the comment here to something like: "final step: we have not seen this host certificate before, prompt the user, then add to known hosts".

On my first pass I was a little confused why we were always asking the user in the final step when they had a certificate. I missed the `return nil` a few lines up because typically you return an error in the body and `return nil` (success) at the end.

russjones replied:
Actually it appears we now do always prompt users if they trust the host?

kontsevoy replied:
added the comments, fixed the 3rd party agent-stored keys issue

russjones commented on lib/client/keyagent.go:
```diff
@@ -199,38 +207,57 @@ func (a *LocalKeyAgent) AddHostSignersToCache(hostSigners []services.CertAuthori
+		if !a.noHosts[host] {
+			fmt.Printf("The authenticity of host '%s' can't be established. "+
+				"Its public key is:\n%s\nAre you sure you want to continue (yes/no)? ",
+				host, ssh.MarshalAuthorizedKey(key))
```
I'm curious why you went with `authorized_keys` format here over `known_hosts` format? Also why not a key fingerprint like OpenSSH does?

russjones commented on lib/client/api.go:
```diff
@@ -953,9 +956,12 @@ func (tc *TeleportClient) ConnectToProxy() (*ProxyClient, error) {
-	if tc.Config.SkipLocalAuth {
-		return nil, trace.BadParameter("failed to authenticate with proxy %v", proxyAddr)
+	// is disabled in configuration, or the user refused connecting to untrusted hosts
+	if tc.Config.SkipLocalAuth || tc.localAgent.UserRefusedHosts() {
```
It might be useful to have these in separate blocks then return different errors. Knowing if authentication failed because because all auth methods were exhausted or because of an invalid host certificate would be useful when debugging issues.

russjones commented on lib/client/keyagent.go:
```diff
@@ -36,6 +36,13 @@ type LocalKeyAgent struct {
+
+	// map of "no hosts". these are hosts that user manually (via keyboard
+	// input) refused connecting to.
+	noHosts map[string]bool
```
When would `noHosts` be used? It doesn't appear to be persisted to disk and isn't `LocalKeyAgent` recreated everytime you run `tsh`?

russjones commented on lib/client/keyagent.go:
```diff
@@ -258,7 +292,7 @@ func (a *LocalKeyAgent) AddKey(host string, username string, key *Key) (*CertAut
 	err := a.keyStore.AddKey(host, username, key)
 	if err != nil {
-		return nil, trace.Wrap(err)
+
```
I think you might have removed this line accidentally?

kontsevoy replied:
fixed

russjones commented on tool/tctl/common/tctl.go:
```diff
@@ -754,39 +758,69 @@ func (a *AuthCommand) GenerateAndSignKeys(client *auth.TunClient) error {
+	//
+	// dump user identity into a single file:
+	//
+	case IdentityFormatFile:
```
I'm not seeing the host certificate being exported and as far as I can tell the `--type` flag does not exist? Because the following from the issue does not work: `tctl auth sign --type=host`.

russjones replied:
Also, we should probably add some documentation for how to use this feature to the Admin Guide.

kontsevoy replied:
`tctl` now dumps the identity to `stdout` by default but you can set `-o` to something else

russjones commented on tool/tctl/common/tctl.go:
```diff
@@ -754,39 +758,69 @@ func (a *AuthCommand) GenerateAndSignKeys(client *auth.TunClient) error {
+			a.output = fmt.Sprintf("%s.pem", a.genUser)
+		}
+		var (
+			output  io.Writer
```
This isn't really consistent with how `tctl` has worked up to this point, `tctl auth export` and `tctl saml export` output to `stdout` and we redirect to a file.

kontsevoy committed:
Code review edits

key-agent changes:

tsh now ignores public keys loaded from the SSH key agent unless they
are certificates. this fixes the problem of 'host auth' callback being
called for 3rd party public keys in the SSH-agent (and tsh had no choice
but to prompt user if he wanted to add thoses hosts to known_hosts file)

tctl auth changes:

`tctl auth` now outputs to stdout by default. If you set -o to a file,
it will use that file instead. Getting rid of -o was not possible
because it's still useful if --format is set to `dir` (OpenSSH
compatible output).

other changes:

- added more detailed comments in a few places
- restored error handling logic
